{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline \n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Binding Specificity by High-Throughput Sequencing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You are studying a protein that binds to DNA. You want to know \"rules\" that the protein uses to recognize DNA it binds to.  You create a library of DNA sequences that looks like the following:\n",
    "\n",
    "`TCGATAACCxxxxxxCCAGACCTCAC`\n",
    "\n",
    "where `x` is any one of the bases `A`, `C`, `T`, or `G`.  You take a sample of this library and then use high-throughput sequencing to characterize it.  These results are in `naive.fastq.gz`.  You then treat the library with your protein, pulling out sequences that bind to your protein.  You sequence the sequences that were pulled out.  These results are in `experimental.fastq.gz`. \n",
    "\n",
    "Write code to read in these libraries.  It should:\n",
    "+ Read the sequences from each library.\n",
    "+ Remove any sequence with a `Q` score for *any* base < 15.\n",
    "+ Make sure that the sequence has the form `TCGATAACCxxxxxxCCAGACCTCAC`.  (Some sequences might have extra bases inserted or deleted).\n",
    "+ Return the counts for each `xxxxxx` sequence within the library."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the enrichment of each base at each position in the `xxxxxx` region.  The enrichment should be a 6x4 matrix, where each row is a position (1-6) and each column is a a base (ACTG).  For each position, enrichment is defined as the change in the relative frequency of a base at that position. For example, the enrichment for base `A` at site 1 is:\n",
    "\n",
    "$$E_{A,1} = \\frac{f_{A,1,experimental}}{f_{A,1,naive}}$$\n",
    "\n",
    "where $f_{A,1} = \\frac{C_{A,1}}{C_{all,1}}$.  $C_{A,1}$ is the number of times $A$ is seen at site one; $C_{all,1}$ is the number of times *any* base is seen at site 1.  \n",
    "\n",
    "According to your analysis, what are the *rules* for recognition at sites 1-6?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "# Dictionary mapping ascii value to Q-score\n",
    "Q_dict = {}\n",
    "for i in range(33,76):\n",
    "    Q_dict[chr(i)] = i-33\n",
    "\n",
    "def q_score_check(qual_string,reject_cutoff=15):\n",
    "    \"\"\"\n",
    "    Check quality scocre\n",
    "    \"\"\"\n",
    "    \n",
    "    for s in qual_string:\n",
    "        if Q_dict[s] < reject_cutoff:\n",
    "            return False\n",
    "        \n",
    "    return True\n",
    "\n",
    "def qual_check(sequence):\n",
    "    \"\"\"\n",
    "    Make sure sequence has the correct form.\n",
    "    \"\"\"\n",
    "    \n",
    "    if sequence[0:9] == \"TCGATAACC\" and sequence[15:] == \"CCAGACCTCAC\":\n",
    "        return True\n",
    "    \n",
    "    return False\n",
    " \n",
    "def read_fastq_gz(file_name):\n",
    "    \"\"\"\n",
    "    Read and process fastq_gz file.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Read sequences and quality scores from the file\n",
    "    seqs = []\n",
    "    quality = []\n",
    "    line_counter = 0\n",
    "    with gzip.open(file_name) as f:\n",
    "        for l in f:\n",
    "            l_ascii = l.decode(\"ascii\").strip()\n",
    "\n",
    "            if l_ascii[0] == \"@\":\n",
    "                line_counter = 1\n",
    "                continue\n",
    "\n",
    "            if line_counter == 1:\n",
    "                seqs.append(l_ascii)\n",
    "                line_counter += 1 \n",
    "            elif line_counter == 2:\n",
    "                line_counter += 1\n",
    "            elif line_counter == 3: \n",
    "                quality.append(l_ascii)\n",
    "                line_counter = 0\n",
    "    \n",
    "    # Count how often each high-quality sequence is seen\n",
    "    total_counts = 0\n",
    "    final = {}\n",
    "    for i in range(len(seqs)):\n",
    "        if q_score_check(quality[i]) and qual_check(seqs[i]):\n",
    "            \n",
    "            s = seqs[i][9:15]\n",
    "            try:\n",
    "                final[s] += 1\n",
    "            except KeyError:\n",
    "                final[s] = 1\n",
    "                \n",
    "            total_counts += 1\n",
    "\n",
    "    # Convert to counts of each base at each site\n",
    "    count_dict = {\"A\":0,\"C\":1,\"G\":2,\"T\":3}\n",
    "    counts = np.zeros((6,4),dtype=np.float)\n",
    "    for k in final.keys():\n",
    "        for i, letter in enumerate(k):\n",
    "            counts[i,count_dict[letter]] += final[k]/total_counts\n",
    "            \n",
    "    return counts\n",
    "\n",
    "# Load and caculate enrichment\n",
    "exp_freqs = read_fastq_gz(\"experimental.fastq.gz\")\n",
    "naive_freqs = read_fastq_gz(\"naive.fastq.gz\")\n",
    "E = exp_freqs/naive_freqs\n",
    "\n",
    "print(E)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+ Site 0: No preference for base\n",
    "+ Site 1: Strongly prefer T \n",
    "+ Site 2: No preference for base\n",
    "+ Site 3: Any base except A\n",
    "+ Site 4: Bases C and T\n",
    "+ Site 5: Strong preference for T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code used to create libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class LibGen:\n",
    "    \n",
    "    def __init__(self,template=\"TCGATAACC{}CCAGACCTCAC\"):\n",
    "        \"\"\"\n",
    "        Generate a fastq file that looks like a sequencing run.\n",
    "        \"\"\"\n",
    "        \n",
    "        self._template = template[:]\n",
    "        self._create_model()\n",
    "    \n",
    "    def _create_model(self):\n",
    "        \"\"\"\n",
    "        The array self._model should be an N x 4 array, where N\n",
    "        is the number of sites to generate.  Each row should add\n",
    "        to 1.0. The row encodes the relative probabilities of\n",
    "        A,C,G, and T at that site. \n",
    "        \"\"\"\n",
    "        \n",
    "        model = np.ones((1,4),dtype=np.float)\n",
    "        self._model = (model.T/np.sum(model,axis=1)).T \n",
    "        \n",
    "        \n",
    "    def create_sample(self,\n",
    "                      q_slope_param=(0,0.2),\n",
    "                      q_intercept_param=(25,1),\n",
    "                      q_noise_param=(0,0.5),\n",
    "                      indel_rate=0.001):\n",
    "        \"\"\"\n",
    "        Generate a random sample.  The quality score decays linearly\n",
    "        with sequence length.  Slope and intercepts are drawn from\n",
    "        distributions encoded by q_slope_param and q_intercept_param.\n",
    "        These are tuples of mean and standard deivation.  q_norm_param\n",
    "        encodes random noise to be added to each quality score.  The\n",
    "        indel_rate is the rate at which random indels will be \n",
    "        incorporated. \n",
    "        \"\"\"\n",
    "        \n",
    "                \n",
    "        # Generate a sequence according to self._model\n",
    "        tmp = []\n",
    "        for i in range(self._model.shape[0]):\n",
    "            tmp.append(np.random.choice([\"A\",\"C\",\"G\",\"T\"],p=self._model[i,:]))\n",
    "        sequence = list(self._template.format(\"\".join(tmp)))\n",
    "        \n",
    "        # Add random indels\n",
    "        if np.random.random() < indel_rate:\n",
    "            index = np.random.choice(range(len(sequence)))\n",
    "            if np.random.random() > 0.5:\n",
    "                sequence.insert(index,np.random.choice([\"A\",\"T\",\"G\",\"C\"]))\n",
    "                sequence = sequence[:-1]\n",
    "            else:\n",
    "                sequence.pop(index)\n",
    "                sequence.append(np.random.choice([\"A\",\"T\",\"G\",\"C\"]))\n",
    "                    \n",
    "        # Generate Q scores\n",
    "        q_slope = -np.abs(np.random.normal(q_slope_param[0],q_slope_param[1]))\n",
    "        q_intercept = np.random.normal(q_intercept_param[0],q_intercept_param[1])\n",
    "        q_noise = np.random.normal(q_noise_param[0],q_noise_param[1],len(sequence))\n",
    "        Q = np.round(np.arange(len(sequence))*q_slope + q_intercept + q_noise,0)\n",
    "        \n",
    "        # Now add noise accoring to Q-scores\n",
    "        p = 10**(-Q/10.)\n",
    "        for i in range(len(sequence)):\n",
    "            if np.random.random() < p[i]:\n",
    "                possibilities = [\"A\",\"C\",\"G\",\"T\"]\n",
    "                possibilities.remove(sequence[i])\n",
    "                sequence[i] = np.random.choice(possibilities)\n",
    "                \n",
    "        seq_string = \"\".join(sequence)\n",
    "        Q_string = \"\".join([chr(int(q)+33) for q in Q])\n",
    "        \n",
    "        # Hack to make sure quality score does not start with \"@\" to avoid breaking \n",
    "        # student parsers...\n",
    "        if Q_string[0] == \"@\":\n",
    "            out = self.create_sample(q_slope_param,\n",
    "                                     q_intercept_param,\n",
    "                                     q_noise_param,\n",
    "                                     indel_rate)\n",
    "        else:\n",
    "            out = \"@{}\\n{}\\n+\\n{}\\n\".format(int(np.random.random()*20000),seq_string,Q_string)\n",
    "        \n",
    "        return out\n",
    "\n",
    "class NaiveLib(LibGen):\n",
    "    \"\"\"\n",
    "    Naive library with a hot \"C\" in position 3, the rest even.\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    def _create_model(self):\n",
    "        \n",
    "        # ACGT\n",
    "        model = np.ones((6,4),dtype=np.float)\n",
    "        \n",
    "        # Make third site \"C\" hot, but others allowed\n",
    "        model[2,1] = 10000.0\n",
    "        \n",
    "        self._model = (model.T/np.sum(model,axis=1)).T       \n",
    "    \n",
    "class ExperimentalLib(LibGen):\n",
    "    \"\"\"\n",
    "    Experimental library with all kinds of interesting single-site rules.\n",
    "    \"\"\"\n",
    "    \n",
    "    def _create_model(self):\n",
    "        \n",
    "        # ACGT\n",
    "        model = np.ones((6,4),dtype=np.float)\n",
    "        \n",
    "        # First site is even probs\n",
    "        \n",
    "        # Second site can *only* by T\n",
    "        model[1,0] = 0.0\n",
    "        model[1,1] = 0.0\n",
    "        model[1,2] = 0.0\n",
    "        model[1,3] = 100.0\n",
    "        \n",
    "        # Make third site \"C\" hot, but others allowed\n",
    "        model[2,1] = 10000.0\n",
    "        \n",
    "        # Make fourth site \"A\" non-existance\n",
    "        model[3,0] = 0.0\n",
    "        \n",
    "        # Make fourth site neither A nor G\n",
    "        model[4,0] = 0.0\n",
    "        model[4,2] = 0.0\n",
    "        \n",
    "        # Make sixth site prefer  C > T > A > G\n",
    "        model[5,0] = 2.0\n",
    "        model[5,1] = 10.0\n",
    "        model[5,2] = 1.0\n",
    "        model[5,3] = 5000.0\n",
    "        \n",
    "        self._model = (model.T/np.sum(model,axis=1)).T    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "f = gzip.open(\"experimental.fastq.gz\",\"wb\")\n",
    "a = ExperimentalLib()\n",
    "for i in range(1000380):\n",
    "    f.write(str.encode(a.create_sample()))\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "f = gzip.open(\"naive.fastq.gz\",\"wb\")\n",
    "a = NaiveLib()\n",
    "for i in range(2321080):\n",
    "    f.write(str.encode(a.create_sample()))\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
